### DISCUSSION TITLE
How to hide the Menu?

### DISCUSSION THREAD

**Discussion by forntoh** [forntoh/LcdMenu#12] [Q&A] [answered]
From @carloslema
_I am stuck, and this is very silly. I have not figured out how to remove the menu once is created. I can initialize it with a switch (splitting LCD int & creating the menu obj.)_

_I tried a few things: from simple `lcd->clear()`, `MenuItem noMenu[] = {}` to brute force by deleteting the pointer to the menu `(delete[] this->currentMenuTable; / delete &this->currentMenuTable;)` bad idea (crahsed)._

**Comment by forntoh** [comment]
@carloslema Do you want to remove the menu and create a new one? Or you just want to hide it temporarily?

**Reply by carloslema** [reply]
Hide main menu & children (plus arrows) temporarily — there should only be one menu.

**Reply by forntoh** [reply]
Should it be like a timeout?, _(i.e. after some time the screen is cleared and backlight is switched off, until the user presses a button or something happens to wake it up)_

**Reply by carloslema** [reply]
@forntoh

Take a look at quick demo I put together — see attached.

As you can see, there is a "**Display State**" — NO menu. The device gives the user feedback, in this case, what button was pressed.

When the user presses the Encoder switch, the device goes into the "**Edit State**" where it displays the MAIN menu (Root Level). In this case, I have several levels deep menu system performing a bunch of actions — updating a local sqlite3 DB, setting up connections etc.

Depending on what function is invoked, the Encoder changes state, from Scrolling up and down the menu, to increase or decrease a parameter value (int), a String which you have with the setText() or panning left or right to scroll horizontally to update a String, like in a password or label (I am working on this).

Also, the Encoder Switch reacts to two types of presses — LONG and SHORT. Mind you, I have not figured out all the stuff as of yet. I am still earning C++, I come from the web dev(HTML, CSS, Vue.js, Node.js and Google API).

All of these actions are USER Initiated — meaning no TimeOut.

But, there are cases where the device should be able to ask the user to perform an action or display a status — which the displayNotification() does well.

Once user is done, they should be able to **exit the "Edit State" to go back to the "Display State"**

I hope this helps. Thank you again for taking a look a look.

Happy Holidays!

\\ carlos
https://user-images.githubusercontent.com/10660649/102924616-efc24400-445f-11eb-92ca-336f9e0b7b0a.mp4

**Reply by agreen72** [reply]
I could use a solution to this as well.  I am trying to edit the String value parameter of an ItemInput object, character by character, with a joystick, before submitting it to menu.setText(value);  to update it.  

There is currently no getValue or getText function exposed to LCDMenu object for external use, those would be useful to start editing the string at it's current state.  Also, tracking the string with the lcd cursor (while blinking) while editing would be nice.  Left/Right on the joystick to move between which digit in the String to edit, and up/down to increment/decrement the character ASCII value at that digit in the String.  Then click to return to menu controls, save and update the value.  I'm sure this can be done now, but it's a bit of a headache and having the user be able to see the current state of the string would be nice.

**Reply by forntoh** [reply]
Unfortunately @agreen72, I don't have time to work on new features for this library.
I will probably start in May 🤷‍♂️

I'm open to PRs, i.e if you can work on some of the features

**Reply by agreen72** [reply]
@carloslema I do not know if you are still working on this project, but what you want can be accomplished by using

If you want it timed you can put it in a while loop after setting a start timer.


```
int debounceDelay = 500; // screen will update every half second (500ms)
int lastDebounce = 0;
int duration = 10 * 1000;  //ten second timer, in milliseconds
int startTime = millis();  //record when loop starts

while ((millis() - startTime) < duration) {

// put your menu printing here, you may have to debounce it so the printspeed of the lcd screen can keep up
if ((millis() - lastDebounce) > debounceDelay) {

// do stuff here
menu.lcd->clear();
menu.lcd->print("first row here");
menu.lcd->setCursor(0,1);
menu.lcd->print("second row here");
}
}
```
getting it to toggle on button press you would just set a boolean flag when entering the display state, and make your while loop dependent on that flag.  Then put your listener code for your input device into the while loop and to exit your display mode, you would have it flip the bool flag back to false or LOW on a button press or something.

**Reply by carloslema** [reply]
@agreen72 thanks for the pointer. I am "sort of" still working on this project. I got my head around @forntoh awesome library and started adding new classes — but have not shared my repo yet. 

You mentioned an "editing the string" function, I assume a use case would be editing a string or password (validation might be an issue). Give me a a day or two to upload my WIP repo — it has few new things that might be useful to you.

About the printing to the LCD... let me take a look and implement what you suggested — the ticket here, that it needs to be a one time screen update regardless if you are on display or menu mode. While I was figuring this out, I ran into the Character LCD's limitations — not being able to display other information regardless of page, like battery level, BLE connection status, etc — plus it's a battery drainer and it needs a buck convertor 3v3 to 5v. That is when I started looking at TFT screen (Nextion LCD, Makerfabs etc) as an alternative.

**Reply by forntoh** [reply]
@carloslema Any progress with the WIP?

I added `backspace()` and `type(character)` functions to remove a character and add a character respectively

**Comment by forntoh** [comment]
I think, adding a new field to `LcdMenu.h` could help you with this.

e.g `bool enableUpdate = true`

_This field will be checked before calling menu `paint()` and before executing callbacks_

Whenever you want to hide the menu to show other information,
you can call `menu.lcd->clear()` followed by `menu.disableUpdate()` then displaying your info, and later reenabling update by calling `menu.enableUpdate()`

**Reply by carloslema** [reply]
Hi @forntoh,

Thanks for the update and keeping this project going. I will take a look over the weekend, I had to put this project on hold, paying projects take priority. I added a few things as well.. such as able to pick from an array (selection menu type of thing). If I  recall correctly, I was wrestling with the battery consumption — I am using a 2K mAh battery that lasts for about 2-3 hours

**Comment by forntoh** [comment, marked as answer]
Now there is a `hide` and a `show` function, calling hide will clear the display.

> It also sets up the stage for #15 ☺
